server/src/test/scala/org/http4s/server/middleware/ThrottleSpec.scala

```diff
@@ -0,0 +1,143 @@
+package org.http4s.server.middleware
+
+import cats.effect.{IO, Timer}
+import cats.effect.laws.util.TestContext
+import org.http4s.{Http4sSpec, HttpApp, Request, Status}
+import cats.implicits._
+import org.http4s.dsl.io._
+import cats.effect.IO.ioEffect
+import org.specs2.matcher.FutureMatchers
+import scala.concurrent.duration._
+import org.specs2.concurrent.ExecutionEnv
+import Throttle._
+
+class ThrottleSpec(implicit ee: ExecutionEnv) extends Http4sSpec with FutureMatchers {
+  "LocalTokenBucket" should {
+
+    "contain initial number of tokens equal to specified capacity" in {
+      val ctx = TestContext()
+      val testTimer: Timer[IO] = ctx.timer[IO]
+
+      val someRefillTime = 1234.milliseconds
+      val capacity = 5
+      val createBucket =
+        TokenBucket.local[IO](capacity, someRefillTime)(ioEffect, testTimer.clock)
+
+      val takeExtraToken = createBucket
+        .flatMap(testee => {
+          val takeFiveTokens: IO[List[TokenAvailability]] =
+            (1 to 5).toList.traverse(_ => testee.takeToken)
+          val checkTokensUpToCapacity =
+            takeFiveTokens.map(tokens =>
+              tokens must contain(TokenAvailable: TokenAvailability).forall)
+          checkTokensUpToCapacity *> testee.takeToken
+        })
+
+      val result = takeExtraToken.unsafeToFuture()
+
+      result must haveClass[TokenUnavailable].await
+    }
+
+    "add another token at specified interval when not at capacity" in {
+      val ctx = TestContext()
+      val testTimer: Timer[IO] = ctx.timer[IO]
+
+      val capacity = 1
+      val createBucket =
+        TokenBucket.local[IO](capacity, 100.milliseconds)(ioEffect, testTimer.clock)
+
+      val takeTokenAfterRefill = createBucket.flatMap(testee => {
+        testee.takeToken *> testTimer.sleep(101.milliseconds) *> testee.takeToken
+      })
+
+      val result = takeTokenAfterRefill.unsafeToFuture()
+
+      ctx.tick(101.milliseconds)
+
+      result must beEqualTo(TokenAvailable).await
+    }
+
+    "not add another token at specified interval when at capacity" in {
+      val ctx = TestContext()
+      val testTimer: Timer[IO] = ctx.timer[IO]
+      val capacity = 5
+      val createBucket =
+        TokenBucket.local[IO](capacity, 100.milliseconds)(ioEffect, testTimer.clock)
+
+      val takeExtraToken = createBucket.flatMap(testee => {
+        val takeFiveTokens: IO[List[TokenAvailability]] = (1 to 5).toList.traverse(_ => {
+          testee.takeToken
+        })
+        testTimer.sleep(300.milliseconds) >> takeFiveTokens >> testee.takeToken
```

Comment:
why does it sleep 300 millis, take 5 tokens then take another token? 

I might be misreading: This test aims to make sure that if trying to take a token at capacity, it won't add another token until demand goes down?

Reply:
No, the intention of the test was to ensure the bucket never adds tokens beyond its capacity, ie we set it to capacity, wait a few refill intervals, then check that there are still only `capacity` tokens.

Reply:
My mistake :+1: